### ISSUE TITLE
🦠 Model Request: Tox21 SR-HSE

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#456]
### Model Title

Tox21 SR-HSE (TDC Commons)

### Publication

Hello @aldeav!

As part of your Outreachy contribution, we have assigned you the dataset "Tox21 SR-HSE" from the Therapeutics Data Commons to try and build a binary classification ML model.
Please copy the provided Google Colab template and use this issue to provide updates on the progress. We'll value not only being able to build the model but also interpreting its results.

**Comment by aldeav** [comment]
@GemmaTuron Thank you!

**Comment by aldeav** [comment]
Hi @GemmaTuron, kindly find the colab notebook attached [here](https://colab.research.google.com/drive/12bjVWKEDnxsKm8wNsMLpQFMvW0H2ZDPg?usp=sharing)

**Comment by aldeav** [comment]
**SUMMARY OF THE TASK**

**1. LOADING DATA**
The dataset I have been assigned is Tox21 SR-HSE. 

Tox21 (Toxicology in the 21st Century) is a dataset/data challenge aimed at predicting toxicity in humans from 7,831 compounds on 12 different targets. ‘SR’ corresponds to stress response signalling pathway. 

The assay ‘SR-HSE’ corresponds to heat shock factor response element. 
A number of chemicals, environmental and physiological stress conditions may lead to the activation of heat shock response and this can cause cellular stress. This toxicity in turn can lead to apoptosis or death of cells. The ‘SR-HSE’ provides the identification of compounds that activate heat shock response signalling in humans. [[1]](https://pubchem.ncbi.nlm.nih.gov/bioassay/743228#section=Data-Table) [[2]](https://arxiv.org/pdf/1503.01445.pdf)
	
In the dataset, an ‘Active’ compound (one that activates heat shock response signalling) is labelled as ‘1’  whereas an ‘Inactive’ compound is labelled as ‘0’. 

The dataset consists of 6467 unique drugs. It is further divided into train, validation and test sets with 4527, 647 and 1293 drugs, respectively. The number of active and inactive molecules in the three sets are as follows:

Train dataset: 
Active Molecules: 270
Inactive Molecules: 4257

Validation dataset: 
Active Molecules: 33
Inactive Molecules: 614

Test dataset: 
Active molecules: 69
Inactive Molecules: 1224

SR-HSE is imbalanced and only 5.946% of the total train data points are active molecules. Such a distribution will make it hard for an estimator to efficiently classify the data points. 

The bar chart plots of inactive and active drugs in the three datasets are as follows:

![Train_data](https://user-images.githubusercontent.com/42399513/198959147-e7075952-12fc-4e7b-a024-680d334c1fe5.png)

![Val_and_Test_data](https://user-images.githubusercontent.com/42399513/198959199-23361dec-35e7-485f-8483-57c1845206dc.png)

**Comment by aldeav** [comment]
**2. ANALYSING DATA**
To draw a few active and inactive molecules, RDKit was used to first convert smiles (a chemical notation that represents a chemical structure that can be used by the computer) to RDKit molecules, which were then used to draw the molecules. 

Two active molecules from Train set:
![Active_mols](https://user-images.githubusercontent.com/42399513/198959620-97cd5439-b154-4aff-ae7d-4dd12cd38ca8.png)


Two inactive molecules from Train set:
![Inactive_mols](https://user-images.githubusercontent.com/42399513/198959665-48227a31-f4f3-486c-8dee-042e9de90fd6.png)


**3. TRAINING THE MODEL** 
Lazy QSAR was installed from github and MorganBinaryClassifier was imported. Since the MorganBinaryClassifier implicitly converts smiles into vectors, the input provided to the classifier was in smiles format. 

After experimenting with various estimators and different time budgets, lrl1 (Logistic Regression with L1 regularisation, best error of 0.1382) classifier was used to fit the training data.

**Comment by aldeav** [comment]
**4. EVALUATION OF THE MODEL**

**_Validation data:_**
**ROC Curve :**
The ROC curve plots the True Positive rate and False Positive rate obtained. 

The True positive rate (TPR) gives the proportion of correct predictions in predictions of positive class
TPR = TP/(TP+FN)

The False positive rate (FPR) gives the proportion of incorrect predictions in positive class.
FPR = FP / (FP + TN)

The more the ROC curve is near the upper left corner (ideally TPR ~= 1 and FPR ~= 0), the better the model is. The ROC curve plotted TPR vs FPR for 145 thresholds. Eventually, the optimal threshold was chosen through Youden’s J statistic.

![Val_ROC](https://user-images.githubusercontent.com/42399513/198959855-8f8f5f39-b404-44f1-b237-f52f75396d23.png)

**Area under the ROC Curve AUC:** 0.764

**CONFUSION MATRIX** (threshold - 0.061):

TP = Active molecules predicted Active
FN = Active molecules predicted Inactive
TN = Inactive molecules predicted Inactive
FP = Inactive molecules predicted Active

![confusion_matrix_val_0 061](https://user-images.githubusercontent.com/42399513/198960122-76cd68e4-ecdd-47aa-a97d-2da180b17462.png)


**Precision**: 0.115
**Recall:** 0.667
**F1 Score:** 0.196
**Accuracy:** 0.72

**Comment by aldeav** [comment]
**_Test data:_**
**ROC Curve:**
![Test_ROC](https://user-images.githubusercontent.com/42399513/198960542-27a2147e-8747-4b5a-9cf9-674683ddcab1.png)


**Area under the ROC Curve AUC:** 0.729

**CONFUSION MATRIX (threshold - 0.061):**
![confusion_matrix_test_0 061](https://user-images.githubusercontent.com/42399513/198960587-0aa5a5f7-d8de-4c20-ab03-4b4049f4c7eb.png)


**Precision:** 0.105
**Recall:** 0.667
**F1 Score:** 0.181
**Accuracy:** 0.677

**Comment by aldeav** [comment]
**5. OBSERVATIONS AND CONCLUSION:**

The SR-HSE dataset is imbalanced with Active Molecules making approximately 6% percent of trained dataset. Such a distribution made it hard for the estimator to efficiently classify the data points. **Two ways to deal with the same:**


1. Since our target label is the active molecule, we can use **class weights** while training which provides more weight to active molecules. 
2. Another method to improve the performance is to **adjust the threshold** of the classifier in accordance to the evaluation metric we wish to perform well on. 


Secondly, finding an optimal threshold, through the ROC, can be done in many ways but for this task I used **Youden’s J statistic**. A number of other methods can be explored in the future to find a threshold that works the best for our problem statement. [[3]](https://machinelearningmastery.com/threshold-moving-for-imbalanced-classification/)

Next, precision and recall are impacted by our choice of threshold. In this case, increasing the threshold from **0.061** resulted in more active molecules being misclassified as inactive i.e. number of FN increased and number of TP decreased (that is, **recall decreased**) whereas decreasing the threshold affected precision. 

**Threshold 1: 0.055**
FN = 9      (27.3%)
TP = 24    (72.7%)

**Threshold 2: 0.061 (best threshold)**
FN = 11    (33.4%)
TP =  22   (66.6%)
Using the current threshold, we are misclassifying 1/3rd of the total active molecules

**Threshold 1: 0.080**
FN = 15    (45.5%)
TP = 18    (54.5%)

In my opinion, the cost of penalising the classifier for misclassifying active molecules as inactive should be higher than its inability to correctly classify inactive molecules. Thus, initially, the objective of decreasing the number of FN or optimising recall while maintaining a satisfactory precision score should be considered the next step in improving the given classifier.  

Nonetheless, the problem statement can move ahead with three objectives:

**1. Optimising Recall by minimising FN
2. Optimising Precision by minimising FP
3. Optimising F1 score and thus provide a good precision and recall**

**Comment by aldeav** [comment]
@GemmaTuron  I'd be grateful for your valuable feedback and suggestions. Thank you!

**Comment by GemmaTuron** [comment]
Hi @aldeav,

Very good job and reasoning, thanks for the work. Things that could also help would be testing other algorithms, not limiting to logistic regression, and changing the molecular descriptors from morgan to other fingerprints.
But at this point I think you are ready to prepare your final application, please link this issue to your contribution in outreachy as well!
I'll close the issue as completed

**Comment by aldeav** [comment]
@GemmaTuron Thank you for your feedback! I did individually try all the estimators from the estimator_list, giving all of them sufficient time budget, and saved all the models. Eventually, I chose LR given its error and AUC. I have mentioned the same in the summary but I will elaborate further to make it more clear. 

Also, I will try and read about various other fingerprints to be better versed.

Thank you for your time and help!
